103_agent-driven-attestation.md

```diff
@@ -0,0 +1,421 @@
+
+To downgrade a Keylime component to a version prior to the introduction of agent-driven attestation, the component must be configured to operate in pull mode. In that case, the downgrade should be seamless.
+
+Switching between push and pull mode is a separate matter. First, all components which are being switched from one mode to the other must be running a version which supports the new push mode. Then, the user will need to perform the steps below, depending on the component in question:
+
+#### Registrar
+
+The registrar will simultaneously support agents operating in push and pull mode, so it is not necessary to change the mode of the registrar. However, the user may need to review the settings which affect the trust decisions made by the registrar (e.g., by configuring a webhook URI, if that feature is desired).
+
+#### Verifier
+
+To switch a verifier's operating mode, change the appropriate configuration option and restart the verifier. After restarting, all agents which are enrolled with the verifier will be deemed to have failed verification. To restart attestation of the agents, the user will need to change the operating mode of each agent (according the directions below) and then issue a command via the tenant. This is so that the tenant can re-check the trust status of the node's identities before restarting attestation.
+
+#### Agent
+
+To switch an agent's operating mode, again, simply change the agent's configuration and restart the agent. After restarting, the agent will attempt to update its registration at the registrar. If this succeeds and the agent's new operating mode is push, the agent will start sending attestations to the verifier specified in the agent's configuration file. Attestations will be rejected by the verifier until attestations are restarted by issuing the tenant command described above.
+
+### Dependency Requirements
+
+This enhancement proposal requires a relatively significant number of new API endpoints, especially for the verifier. These include those needed to support the new attestation protocol and challenge–response authentication mechanism. Changes to existing endpoints may also be appropriate to expose any new information stored in the database, such as more granular information about the trust status of the individual identities of a node.
+
+It suggested that, just as the Rust agent uses [Actix](https://actix.rs/) to serve its REST API, a similar web micro-framework is added as a dependency to the Python codebase. This will provide a natural way to define routes and request handlers and help newcomers understand how the code is organised.
```

Comment:
what would be the reach of this? like dropping tornado in favor of the new framework?

Reply:
You know what, I was basing my understanding of how KL implements its REST APIs on the registrar's code (which uses http.server from the standard library), assuming it was the same globally. However, I see now that the verifier APIs have been implemented using Tornado. Since KL already has a web framework as a dependency, I'll use that naturally and just refactor the existing code a bit

Reply:
Yeah I think we should keep tornado, but should migrate the registrar also to it.